src/hook.c

```diff
@@ -0,0 +1,153 @@
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "common.h"
+#include "posix.h"
+#include "repository.h"
+
+#include "git2/hook.h"
+
+#define GIT_HOOKS_DIR "hooks/"
+
+const char * const githooks[] = {
+	"applypatch-msg",
+	"pre-applypatch",
+	"post-applypatch",
+	"pre-commit",
+	"prepare-commit-msg",
+	"commit-msg",
+	"post-commit",
+	"pre-rebase",
+	"post-checkout",
+	"post-merge",
+	"pre-push",
+	"pre-receive",
+	"update",
+	"post-receive",
+	"post-update",
+	"push-to-checkout",
+	"pre-auto-gc",
+	"post-rewrite",
+};
+
+static int hook_dir(git_buf *out_dir, git_repository *repo)
+{
+	int err;
+	git_buf tmp_path = GIT_BUF_INIT;
+	git_buf cfg_path = GIT_BUF_INIT;
+	git_config *cfg;
+
+	assert(out_dir && repo);
+
+	/* We need to check for an override in the repo config */
+	err = git_repository_config__weakptr(&cfg, repo);
+	if (err != 0)
+		return err;
+
+	err = git_config_get_path(&cfg_path, cfg, "core.hooksPath");
+	if (err == GIT_ENOTFOUND) {
+		giterr_clear();
+		git_buf_puts(&tmp_path, git_repository_commondir(repo));
+		git_buf_joinpath(&tmp_path, tmp_path.ptr, GIT_HOOKS_DIR);
+	} else if (err == GIT_OK) {
+		git_buf_joinpath(&tmp_path, git_repository_commondir(repo), cfg_path.ptr);
+		git_path_resolve_relative(&tmp_path, 0);
+		git_buf_free(&cfg_path);
+	} else {
+		return -1;
+	}
+
+	*out_dir = tmp_path;
+	return 0;
+}
+
+int git_hook_dir(char **out_dir, git_repository *repo) {
+	git_buf hook_dir_buf = GIT_BUF_INIT;
+	int err;
+
+	assert(out_dir != NULL);
+
+	err = hook_dir(&hook_dir_buf, repo);
+	if (err != 0) {
+		*out_dir = NULL;
+		return -1;
+	}
+
+	*out_dir = git_buf_detach(&hook_dir_buf);
+
+	return 0;
+}
+
+static int build_hook_path(char **out_path, git_repository *repo, const char *hook_name)
+{
+	int err;
+	git_buf hook_path = GIT_BUF_INIT;
+
+	assert(hook_name);
+
+	err = hook_dir(&hook_path, repo);
+	if (err != 0)
+		return -1;
+
+	err = git_buf_joinpath(&hook_path, hook_path.ptr, hook_name);
+	if (err != 0)
+		return -1;
+
+	*out_path = git_buf_detach(&hook_path);
+
+	return 0;
+}
+
+static int check_hook_path(const char *hook_path)
+{
+	int err;
+	struct stat hook_stat;
+
+	assert(hook_path);
+
+	/* Skip missing hooks */
+	err = p_stat(hook_path, &hook_stat);
+	if (err)
+		return -1;
+
+#ifndef GIT_WIN32
+	/* Check exec bits */
+	if ((hook_stat.st_mode & (S_IXUSR|S_IXGRP|S_IXOTH)) == 0) return -1;
+#endif
+
+	return 0;
+}
+
+int git_hook_foreach(
+	git_repository *repo,
+	git_hook_foreach_cb callback,
+	void *payload)
+{
+	size_t hook_id = 0;
+
+	assert(repo && callback);
+
+	for (hook_id = 0; hook_id < ARRAY_SIZE(githooks); hook_id++) {
+		const char *hook_name = githooks[hook_id];
+		int err = 0;
+		char *hook_path = NULL;
+
+		err = build_hook_path(&hook_path, repo, hook_name);
+		if (err != 0)
+			continue;
+
+		err = check_hook_path(hook_path);
+		git__free(hook_path);
+
+		if (err != 0)
+			continue;
+
+		callback(hook_name, payload);
```

Comment:
We usually allow callbacks to abort iteration by returning non-zero. Do we want to do the same here?

Reply:
Fixed.